### PR TITLE
Avoid COW materialization in `at::parallel_for/parallel_reduce`

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119507
* __->__ #120455

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: c10/util/ParallelGuard.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+
+namespace c10 {
+
+// Returns true if executed while a ParallelGuard object is alive in the current
+// thread.
+C10_API bool is_parallel_guard_alive();
```
Comment (albanD) [inline c10/util/ParallelGuard.h:9]: nit: if we want this to be closer to GradMode and other similar flags, you can move this to `ParallelGuard::is_enabled()`.

File: c10/util/ParallelGuard.cpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#include <c10/util/ParallelGuard.h>
+
```
Comment (albanD) [inline c10/util/ParallelGuard.cpp:2]: Not sure which build files need updating for a new cpp file in this folder...

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -446,6 +446,10 @@
   self: grad
   result: auto_linear
 
+- name: _test_parallel_materialize(Tensor self, int num_parallel, bool skip_main_thread=False) -> Tensor
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:449]: You never run backward on it so this can be omitted.

**Review by albanD — COMMENTED** [review]
Very cool!

**Review by albanD — APPROVED** [review]
Sounds good! Thanks!

**Comment by huydhn** [comment]
This is also breaking XLA test, i.e. https://hud.pytorch.org/pytorch/pytorch/commit/2ebf2c88baa4667d55eda92f4c8424db505af781.  The log classifier shows the wrong error there but the crash is at the bottom of the log `test_parallel_cow_materialize_error_xla_bfloat16 (__main__.TestTorchDeviceTypeXLA) ... *** Received signal 11 ***`

**Pull request closed without merging**